File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/Compressed.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.bctags.TAG_COMPRESSED
import com.blockchaincommons.bcur.UR
import com.blockchaincommons.bcur.URCodable
import com.blockchaincommons.dcbor.Cbor
import com.blockchaincommons.dcbor.CborTaggedCodable
import com.blockchaincommons.dcbor.CborTaggedUtils
import com.blockchaincommons.dcbor.Tag
import com.blockchaincommons.dcbor.tagsForValues
import java.util.zip.CRC32
import java.util.zip.Deflater
import java.util.zip.Inflater

/**
 * A compressed binary object with integrity verification.
 *
 * [Compressed] provides efficient storage and transmission of binary data
 * using the raw DEFLATE algorithm (no zlib/gzip wrapper). It includes a
 * CRC32 checksum for integrity verification and an optional cryptographic
 * [Digest].
 *
 * CBOR array format: `[checksum, decompressed_size, compressed_data, digest?]`
 */
class Compressed private constructor(
    /** CRC32 checksum of the decompressed data. */
    val checksum: UInt,
    /** Size of the original decompressed data in bytes. */
    val decompressedSize: Int,
    /** The compressed data (or original data if compression was ineffective). */
    private val compressedData: ByteArray,
    /** Optional cryptographic digest of the content. */
    private val digest: Digest?,
) : DigestProvider,
    CborTaggedCodable,
    URCodable {

    /** The size of the compressed data in bytes. */
    val compressedSize: Int get() = compressedData.size

    /** The compression ratio (compressed / decompressed). Values < 1.0 indicate effective compression. */
    val compressionRatio: Double
        get() = compressedSize.toDouble() / decompressedSize.toDouble()

    /** Returns the optional digest, or `null` if none was associated. */
    fun digestOrNull(): Digest? = digest

    /** Whether this compressed data has an associated digest. */
    val hasDigest: Boolean get() = digest != null

    /**
     * Decompresses and returns the original data.
     *
     * @throws BcComponentsException.Compression if decompression fails or
     *   the checksum does not match.
     */
    fun decompress(): ByteArray {
        if (compressedData.size >= decompressedSize) {
            return compressedData.copyOf()
        }

        val inflater = Inflater(true) // nowrap = true for raw DEFLATE
        inflater.setInput(compressedData)
        val output = ByteArray(decompressedSize)
        try {
            val resultLength = inflater.inflate(output)
            if (resultLength != decompressedSize) {
                throw BcComponentsException.compression("corrupt compressed data")
            }
        } catch (e: java.util.zip.DataFormatException) {
            throw BcComponentsException.compression("corrupt compressed data")
        } finally {
            inflater.end()
        }

        if (computeCrc32(output) != checksum) {
            throw BcComponentsException.compression("compressed data checksum mismatch")
        }

        return output
    }

    // -- DigestProvider --

    /**
     * Returns the cryptographic digest associated with this compressed data.
     *
     * @throws IllegalStateException if no digest was associated at compression
     *   time. Use [digestOrNull] to check for the presence of a digest without
     *   throwing.
     */
    override fun digest(): Digest =
        digest ?: throw IllegalStateException("No digest associated with this Compressed")

    // -- equals / hashCode --

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Compressed) return false
        return checksum == other.checksum &&
            decompressedSize == other.decompressedSize &&
            compressedData.contentEquals(other.compressedData) &&
            digest == other.digest
    }

    override fun hashCode(): Int {
        var result = checksum.hashCode()
        result = 31 * result + decompressedSize
        result = 31 * result + compressedData.contentHashCode()
        result = 31 * result + (digest?.hashCode() ?: 0)
        return result
    }

    // -- toString --

    override fun toString(): String {
        val checksumHex = "%08x".format(checksum.toInt())
        return "Compressed($checksumHex, $compressedSize/$decompressedSize, ${"%.2f".format(compressionRatio)}, ${digest?.shortDescription() ?: "None"})"
    }

    // -- CBOR --

    override fun cborTags(): List<Tag> = tagsForValues(listOf(TAG_COMPRESSED))

    override fun untaggedCbor(): Cbor {
        val elements = mutableListOf(
            Cbor.fromUInt(checksum),
            Cbor.fromInt(decompressedSize),
            Cbor.fromByteString(compressedData),
        )
        if (digest != null) {
            elements.add(digest.taggedCbor())
        }
        return Cbor.fromArray(elements)
    }

    // -- Companion --

    companion object {
        /**
         * Creates a [Compressed] from pre-computed parameters. Low-level
         * constructor for deserialization or pre-compressed data.
         *
         * @throws BcComponentsException.Compression if the compressed data
         *   is larger than the decompressed size.
         */
        fun create(
            checksum: UInt,
            decompressedSize: Int,
            compressedData: ByteArray,
            digest: Digest?,
        ): Compressed {
            if (compressedData.size > decompressedSize) {
                throw BcComponentsException.compression(
                    "compressed data is larger than decompressed size",
                )
            }
            return Compressed(checksum, decompressedSize, compressedData.copyOf(), digest)
        }

        /**
         * Compresses the given [decompressedData] using DEFLATE.
         *
         * If compression would increase the data size (common for small or
         * already-compressed inputs), the original data is stored instead.
         */
        fun fromDecompressedData(
            decompressedData: ByteArray,
            digest: Digest? = null,
        ): Compressed {
            val checksum = computeCrc32(decompressedData)
            val decompressedSize = decompressedData.size

            val deflater = Deflater(Deflater.DEFAULT_COMPRESSION, true) // nowrap = true
            deflater.setInput(decompressedData)
            deflater.finish()
            val buffer = ByteArray(decompressedData.size + 256)
            val compressedLength = deflater.deflate(buffer)
            deflater.end()
            val compressedData = buffer.copyOfRange(0, compressedLength)

            return if (compressedLength > 0 && compressedLength < decompressedSize) {
                Compressed(checksum, decompressedSize, compressedData, digest)
            } else {
                Compressed(checksum, decompressedSize, decompressedData.copyOf(), digest)
            }
        }

        /** Decodes a [Compressed] from untagged CBOR (an array). */
        fun fromUntaggedCbor(cbor: Cbor): Compressed {
            val elements = cbor.tryArray()
            if (elements.size < 3 || elements.size > 4) {
                throw com.blockchaincommons.dcbor.CborException.msg(
                    "invalid number of elements in compressed",
                )
            }
            val checksum = elements[0].tryUInt()
            val decompressedSize = elements[1].tryInt()
            val compressedData = elements[2].tryByteStringData()
            val digest = if (elements.size == 4) {
                Digest.fromTaggedCbor(elements[3])
            } else {
                null
            }
            return create(checksum, decompressedSize, compressedData, digest)
        }

        /** Decodes a [Compressed] from tagged CBOR. */
        fun fromTaggedCbor(cbor: Cbor): Compressed =
            CborTaggedUtils.fromTaggedCbor(
                cbor,
                tagsForValues(listOf(TAG_COMPRESSED)),
            ) { fromUntaggedCbor(it) }

        /** Decodes a [Compressed] from tagged CBOR binary data. */
        fun fromTaggedCborData(data: ByteArray): Compressed =
            CborTaggedUtils.fromTaggedCborData(
                data,
                tagsForValues(listOf(TAG_COMPRESSED)),
            ) { fromUntaggedCbor(it) }

        /** Decodes a [Compressed] from a UR. */
        fun fromUr(ur: UR): Compressed {
            ur.checkType("compressed")
            return fromUntaggedCbor(ur.cbor)
        }

        /** Decodes a [Compressed] from a UR string. */
        fun fromUrString(urString: String): Compressed =
            fromUr(UR.fromUrString(urString))

        private fun computeCrc32(data: ByteArray): UInt {
            val crc = CRC32()
            crc.update(data)
            return crc.value.toUInt()
        }
    }
}

File: python/bc-components/src/bc_components/_compressed.py
"""Compressed binary object with integrity verification.

Uses raw DEFLATE (RFC 1951) compression via the ``zlib`` stdlib module,
with CRC-32 checksum and optional cryptographic digest.
"""

from __future__ import annotations

import zlib

from bc_crypto import crc32
from bc_tags import TAG_COMPRESSED, tags_for_values
from dcbor import CBOR, Tag

from ._digest import Digest
from ._error import BCComponentsError


class Compressed:
    """A compressed binary object with CRC-32 checksum and optional digest."""

    __slots__ = ("_checksum", "_decompressed_size", "_compressed_data", "_digest")

    def __init__(
        self,
        checksum: int,
        decompressed_size: int,
        compressed_data: bytes,
        digest: Digest | None = None,
    ) -> None:
        if len(compressed_data) > decompressed_size:
            raise BCComponentsError.compression(
                "compressed data is larger than decompressed size"
            )
        self._checksum = checksum
        self._decompressed_size = decompressed_size
        self._compressed_data = bytes(compressed_data)
        self._digest = digest

    # --- Construction --------------------------------------------------

    @staticmethod
    def new(
        checksum: int,
        decompressed_size: int,
        compressed_data: bytes,
        digest: Digest | None = None,
    ) -> Compressed:
        """Low-level constructor for pre-compressed data or deserialization."""
        return Compressed(checksum, decompressed_size, compressed_data, digest)

    @staticmethod
    def from_decompressed_data(
        decompressed_data: bytes | bytearray,
        digest: Digest | None = None,
    ) -> Compressed:
        """Compress *decompressed_data* using raw DEFLATE (level 6).

        If compression would increase the size, the original data is stored.
        """
        decompressed_data = bytes(decompressed_data)
        # Raw DEFLATE with wbits=-15, level 6 to match Rust miniz_oxide level 6
        compressed_data = zlib.compress(decompressed_data, level=6, wbits=-15)
        checksum = crc32(decompressed_data)
        decompressed_size = len(decompressed_data)
        compressed_size = len(compressed_data)
        if compressed_size != 0 and compressed_size < decompressed_size:
            return Compressed(checksum, decompressed_size, compressed_data, digest)
        return Compressed(
            checksum, decompressed_size, decompressed_data, digest,
        )

    # --- Decompression -------------------------------------------------

    def decompress(self) -> bytes:
        """Decompress and return the original data, verifying the CRC-32 checksum."""
        compressed_size = len(self._compressed_data)
        if compressed_size >= self._decompressed_size:
            # Data was stored uncompressed
            return self._compressed_data

        try:
            decompressed = zlib.decompress(self._compressed_data, wbits=-15)
        except zlib.error:
            raise BCComponentsError.compression("corrupt compressed data")

        if crc32(decompressed) != self._checksum:
            raise BCComponentsError.compression(
                "compressed data checksum mismatch"
            )
        return decompressed

    # --- Accessors -----------------------------------------------------

    @property
    def checksum(self) -> int:
        return self._checksum

    @property
    def decompressed_size(self) -> int:
        return self._decompressed_size

    @property
    def compressed_size(self) -> int:
        return len(self._compressed_data)

    @property
    def compression_ratio(self) -> float:
        if self._decompressed_size == 0:
            return float("nan")
        return self.compressed_size / self._decompressed_size

    @property
    def digest_opt(self) -> Digest | None:
        return self._digest

    @property
    def has_digest(self) -> bool:
        return self._digest is not None

    def digest(self) -> Digest:
        """Return the digest, raising if none is present."""
        if self._digest is None:
            raise ValueError("Compressed has no digest")
        return self._digest

    # --- CBOR ----------------------------------------------------------

    @staticmethod
    def cbor_tags() -> list[Tag]:
        return tags_for_values([TAG_COMPRESSED])

    def untagged_cbor(self) -> CBOR:
        elements: list[CBOR] = [
            CBOR.from_int(self._checksum),
            CBOR.from_int(self._decompressed_size),
            CBOR.from_bytes(self._compressed_data),
        ]
        if self._digest is not None:
            elements.append(self._digest.tagged_cbor())
        return CBOR.from_array(elements)

    def tagged_cbor(self) -> CBOR:
        tags = self.cbor_tags()
        return CBOR.from_tagged_value(tags[0], self.untagged_cbor())

    def tagged_cbor_data(self) -> bytes:
        return self.tagged_cbor().to_cbor_data()

    @classmethod
    def from_untagged_cbor(cls, cbor: CBOR) -> Compressed:
        elements = cbor.try_array()
        if len(elements) < 3 or len(elements) > 4:
            raise BCComponentsError.invalid_data(
                "compressed", "invalid number of elements"
            )
        checksum = elements[0].try_int()
        decompressed_size = elements[1].try_int()
        compressed_data = elements[2].try_byte_string()
        digest: Digest | None = None
        if len(elements) == 4:
            digest = Digest.from_tagged_cbor(elements[3])
        return cls.new(checksum, decompressed_size, compressed_data, digest)

    @classmethod
    def from_tagged_cbor(cls, cbor: CBOR) -> Compressed:
        tags = cls.cbor_tags()
        item = cbor.try_expected_tagged_value(tags[0])
        return cls.from_untagged_cbor(item)

    @classmethod
    def from_tagged_cbor_data(cls, data: bytes) -> Compressed:
        cbor = CBOR.from_data(data)
        return cls.from_tagged_cbor(cbor)

    # --- Dunder methods ------------------------------------------------

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Compressed):
            return NotImplemented
        return (
            self._checksum == other._checksum
            and self._decompressed_size == other._decompressed_size
            and self._compressed_data == other._compressed_data
            and self._digest == other._digest
        )

    def __hash__(self) -> int:
        return hash(
            (self._checksum, self._decompressed_size, self._compressed_data)
        )

    def __repr__(self) -> str:
        digest_str = (
            self._digest.short_description() if self._digest is not None else "None"
        )
        return (
            f"Compressed(checksum: {self._checksum:08x}, "
            f"size: {self.compressed_size}/{self._decompressed_size}, "
            f"ratio: {self.compression_ratio:.2f}, "
            f"digest: {digest_str})"
        )
